keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md

```diff
@@ -0,0 +1,398 @@
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Provide an interface so that users can implement a custom queue of scheduler.
+With such, users can build their custom queue implementation at will, and plumb into
+the scheduler seamlessly. This can satisfy the business needs such as sophisticated
+pods sorting and internal queuing mechanics.
+
+## Motivation
+
+The current internal queue implementation ([scheduling_queue.go#PriorityQueue](https://github.com/kubernetes/kubernetes/blob/ee459b8969ed2abfed79a07d4ac9d41f13f18ce6/pkg/scheduler/internal/queue/scheduling_queue.go#L126))
+works for most cases, but the limited interfaces exposure makes it hard to
+implement a scheduler plugin for sophisticated requirements.
+
+One requirement in terms of multi-tenancy support:  
+There are many pods from 2 users (userA and userB), and the target is to ensure resource
+usage (E.g. userA gets x CPUs) ratio between userA and userB is 1:1. Sorting the pods by
+the resource usage in `Less` function doesn't work properly, because the resource usage for
+userA/userB will be updated dynamically once a pod is selected and bound, while the
+algorithm (heapsort) that used by the current queue to get the next pods depends on static
+data.
+
+Other requirements come from current semi-exposed functions/parameters:
+For example, the interval of function flushUnschedulableQLeftover, although configurable,
+is immutable after it gets initialized. A requirement is to adjust the interval dynamically
```

Comment:
What is the use case for this?

The intend of the flush is to make sure we are not leaving pods without scheduling retries. Ideally, it shouldn't exist and retries should entirely be based on cluster events. But realistically, it's hard to catch all events.
But depending on your use case, you might just want a mechanism to trigger the retry of *certain* pods.

Reply:
This case is used to show that there are many possible customer cases that we cannot handle with the current internal queue, and we need a solution for customers to implement the queue at their will. This is also the reason that why we want the customer the implement the whole queue.